### PR TITLE
Implemented Gemini (#1490)

### LINKED ISSUE #1490: Gemini Support?
### 🚀 The feature

Add support for Google's Gemini.

### Motivation, pitch

I would like to use this in the ongoing Gemini API developer competition in which I'm making a pretty straight-forward agentic AI for Windows. Upon discovering this, I believed it would be useful to me, but then I realized that there is no support for Gemini. Just wanted to know if there is a reason it isn't out yet? or how long it might take? Is it something that is open for contributions?

### PR DESCRIPTION
**Pull request by into-the-night**
## Description

Added support for Gemini using `google-generativeai` module. (sorry it took so long 👍)

Fixes #1490 

## Type of change

- [✅] New feature (non-breaking change which adds functionality)
- [✅] Documentation update

## How Has This Been Tested?

Also added unit test for Gemini in `tests/llms/test_gemini_llm.py`

Please delete options that are not relevant.

- [✅] Unit Test

## Checklist:

- [✅] My code follows the style guidelines of this project
- [✅] I have performed a self-review of my own code
- [✅] I have commented my code, particularly in hard-to-understand areas
- [✅] I have made corresponding changes to the documentation
- [✅] My changes generate no new warnings
- [✅] I have added tests that prove my fix is effective or that my feature works
- [✅] New and existing unit tests pass locally with my changes
- [?] Any dependent changes have been merged and published in downstream modules
- [✅] I have checked my code and corrected any misspellings

## Maintainer Checklist

- [ ] closes #1490  (Replace xxxx with the GitHub issue number)
- [ ] Made sure Checks passed

### CODE REVIEW

**Comment by into-the-night** [comment]
I have a doubt. I noticed there is already a `Google AI` mentioned in the docs (using `litellm`), should I add `Gemini` as a seperate page to the docs or update the `Google AI` page?

**Comment by Dev-Khant** [comment]
> I have a doubt. I noticed there is already a `Google AI` mentioned in the docs (using `litellm`), should I add `Gemini` as a seperate page to the docs or update the `Google AI` page? @Dev-Khant

Hey Thanks for the PR, you should `litellm gemini` with your new doc page.

File: pyproject.toml [hunk 1]
```diff
@@ -22,6 +22,7 @@ openai = "^1.33.0"
 posthog = "^3.5.0"
 pytz = "^2024.1"
 sqlalchemy = "^2.0.31"
+google-generativeai = "^0.8.3"
```
Comment (Dev-Khant) [inline pyproject.toml:25]: Hey make a optional dependency. You can check how we have done for other LLMs

File: mem0/llms/gemini.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+        Returns:
+            list: The list of tools in the required format.
+        """
+        if tools:
+            new_tools = []
+
+            for tool in tools:
+                func = tool["function"].copy()
+                func["parameters"].pop("additionalProperties", None)
```
Comment (Dev-Khant) [inline mem0/llms/gemini.py:92]: Are you sure we don't have to change the underlying tools dict structure here?

**Review by Dev-Khant — COMMENTED** [review]
Make sure to make changes to `mint.json` for updating doc page.

File: pyproject.toml (hunk 1 above)
Comment (into-the-night) [reply pyproject.toml:25]: Done!

File: mem0/llms/gemini.py (hunk 2 above)
Comment (into-the-night) [reply mem0/llms/gemini.py:92]: oh yea I didn't consider nested object properties in the tools dict. In that case, some more changes would need to be made. I'll do that asap.

File: pyproject.toml [hunk 3]
```diff
@@ -18,7 +18,7 @@ readme = "README.md"
 python = ">=3.9,<4.0"
 qdrant-client = "^1.9.1"
 pydantic = "^2.7.3"
-openai = "^1.33.0"
```
Comment (Dev-Khant) [inline pyproject.toml:21]: Hey, Is there a reason to change this?

Comment (into-the-night) [reply pyproject.toml:21]: Ah I think I know why this happened, actually `google-generativeai` raised an error which was fixed by updating `protobuf` (which updated `openai`) but after I made `google-generativeai` optional, I dowgraded `openai` to the earlier version and that's where this changed.

Comment (Dev-Khant) [reply pyproject.toml:21]: So can you make it this version of higher meaning doing this: `^1.33.0` ?

Comment (into-the-night) [reply pyproject.toml]: Yea, sorry for that 😅. its fixed now.

Comment (Dev-Khant) [reply pyproject.toml]: Thanks!!

**Review by Dev-Khant — COMMENTED** [review]
Make sure that `.lock` file remains same as there is no change in `pyproject.toml`.

**Comment by into-the-night** [comment]
> Make sure that `.lock` file remains same as there is no change in `pyproject.toml`.

I see there have been a few changes in the `poetry.lock` file as some packages have been updated, this is poetry's doing. When you say `poetry add openai=^x.xx` it shows `openai ="^x.xx"` in the `pyproject.toml` but actually installs the `latest` version of the package. Its fixed now.

**Comment by Dev-Khant** [comment]
Hey @into-the-night `pyproject.toml` still looks odd. Can you please check and remove changes from `.toml` and `.lock` files if not necessary.

**Comment by into-the-night** [comment]
Wow that was pretty sneaky but something new to learn about line-endings

**Review by Dev-Khant — APPROVED** [review]

**Comment by Dev-Khant** [comment]
LGTM!! Thanks @into-the-night for the quick resolution and for the contribution!

**Comment by into-the-night** [comment]
Thank you so much for your patience😁 we go up brick by brick!

**Pull request merged**